engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/StraightThroughProcessingLoopValidator.java

```diff
@@ -0,0 +1,222 @@
+   *
+   * @param resource the resource file that's getting validated
+   * @param executableProcesses the list of processes in this resource
+   * @return an Either of which the left contains the failure message if any loops have been found
+   */
+  public static Either<Failure, ?> validate(
+      final DeploymentResource resource, final List<ExecutableProcess> executableProcesses) {
+
+    final List<Failure> failures = new ArrayList<>();
+    for (final ExecutableProcess executableProcess : executableProcesses) {
+      hasStraightThroughProcessingLoop(resource, executableProcess).ifLeft(failures::add);
+    }
+
+    if (failures.isEmpty()) {
+      return Either.right(null);
+    } else {
+      final StringWriter writer = new StringWriter();
+      failures.forEach(failure -> writer.write(failure.getMessage()));
+      return Either.left(new Failure(writer.toString()));
+    }
+  }
+
+  /**
+   * Checks a given process for any straight-through processing loops.
+   *
+   * <p>Note that once we've found a loop we don't keep searching to see if there are more!
+   *
+   * @param resource the resource file that's getting validated
+   * @param executableProcess the process we are validating
+   * @return an Either of which the left contains the failure message if any loops have been found
+   */
+  private static Either<Failure, ?> hasStraightThroughProcessingLoop(
+      final DeploymentResource resource, final ExecutableProcess executableProcess) {
+    final var straightThroughElements = getStraightThroughElementsInProcess(executableProcess);
+
+    for (final ExecutableFlowNode element : straightThroughElements) {
+      final var potentialLoop = new LinkedList<ExecutableFlowNode>();
+      final var result = checkForStraightThroughProcessingLoop(potentialLoop, element);
+      if (result.isLeft()) {
+        final String failureMessage =
+            createFailureMessage(resource, executableProcess, result.getLeft());
+        return Either.left(new Failure(failureMessage));
+      }
+    }
+
+    return Either.right(null);
+  }
+
+  /**
+   * When looking for straight-through processing loops it is not necessary to check in all places
+   * of the process. We only have to check the path of straight-through elements. This method
+   * returns a list of all the straight-through elements in a process.
+   *
+   * @param executableProcess the process
+   * @return a list of all straight-through elements in the given process
+   */
+  private static List<ExecutableFlowNode> getStraightThroughElementsInProcess(
+      final ExecutableProcess executableProcess) {
+
+    return executableProcess.getFlowElements().stream()
+        .filter(
+            flowElement ->
+                STRAIGHT_THROUGH_PROCESSING_ELEMENT_TYPES.contains(flowElement.getElementType())
+                    && !WHITELISTED_ELEMENT_TYPES.contains(flowElement.getElementType()))
+        .map(ExecutableFlowNode.class::cast)
+        .toList();
+  }
```

Comment:
:question: I'm wondering whether it would find an issue when we have first an service or user tasks and then such loops?

Reply:
:question: Wouldn't that also mean it would detect a loop - even if between are waitstate is. Do we also want to reject this one? Like you have  task -> user task -> task. This method would return task, task and if last task connects to first task it would detect the loop right?

Reply:
Ok looks like I'm able to deploy this:

```

                Bpmn.createExecutableProcess(processId)
                    .startEvent()
                    .task("task1")
                    .userTask("test")
                    .task("task2")
                    .connectTo("task1")
                    .done()

```

Reply:
Can we add another test that this works and is not rejected?

Reply:
And that this is rejected again:

```

  @Test
  public void shouldRejectDeploymentWithSimpleUndefinedTaskLoop() {
    // given
    final var processId = Strings.newRandomValidBpmnId();

    // when
    final var rejectedDeployment =
        ENGINE
            .deployment()
            .withXmlResource(
                Bpmn.createExecutableProcess(processId)
                    .startEvent()
                    .userTask("test")
                    .task("task1")
                    .task("task2")
                    .connectTo("task1")
                    .done())
            .expectRejection()
            .deploy();

    // then
    Assertions.assertThat(rejectedDeployment)
        .hasKey(ExecuteCommandResponseDecoder.keyNullValue())
        .hasRecordType(RecordType.COMMAND_REJECTION)
        .hasIntent(DeploymentIntent.CREATE)
        .hasRejectionType(RejectionType.INVALID_ARGUMENT);
    assertThat(rejectedDeployment.getRejectionReason())
        .contains(String.format("Process: %s", processId))
        .contains(GENERIC_REJECTION_MESSAGE + "task1 > task2 > task1");
  }

```

Reply:
The following is not rejected:


```

  @Test
  public void shouldRejectDeploymentWithSimpleUndefinedTaskLoop() {
    // given
    final var processId = Strings.newRandomValidBpmnId();

    // when
    final var rejectedDeployment =
        ENGINE
            .deployment()
            .withXmlResource(
                Bpmn.createExecutableProcess(processId)
                    .startEvent()
                    .task("task1")
                    .subProcess(
                        "test",
                        subProcessBuilder ->
                            ((SubProcessBuilder) subProcessBuilder)
                                .embeddedSubProcess()
                                .startEvent()
                                .task("task2")
                                .endEvent())
                    .task("task3")
                    .connectTo("task1")
                    .done())
            .expectRejection()
            .deploy();

    // then
    Assertions.assertThat(rejectedDeployment)
        .hasKey(ExecuteCommandResponseDecoder.keyNullValue())
        .hasRecordType(RecordType.COMMAND_REJECTION)
        .hasIntent(DeploymentIntent.CREATE)
        .hasRejectionType(RejectionType.INVALID_ARGUMENT);
    assertThat(rejectedDeployment.getRejectionReason())
        .contains(String.format("Process: %s", processId))
        .contains(GENERIC_REJECTION_MESSAGE + "task1 > task2 > task1");
  }

```


So embedded sub processes and start events I guess we also need to consider right?